File: app/src/main/java/com/abishov/hexocat/home/repository/RepositoryAdapter.java
package com.abishov.hexocat.home.repository;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.abishov.hexocat.R;
import com.abishov.hexocat.home.repository.RepositoryAdapter.RepositoryViewHolder;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.functions.Consumer;

public final class RepositoryAdapter extends Adapter<RepositoryViewHolder>
        implements Consumer<List<RepositoryViewModel>> {
    private final LayoutInflater layoutInflater;
    private final List<RepositoryViewModel> repositories;
    private final TrendingViewClickListener trendingViewClickListener;

    public RepositoryAdapter(LayoutInflater layoutInflater, TrendingViewClickListener clickListener) {
        this.layoutInflater = layoutInflater;
        this.trendingViewClickListener = clickListener;
        this.repositories = new ArrayList<>();
    }

    @Override
    public RepositoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new RepositoryViewHolder((RepositoryItemView) layoutInflater.inflate(
                R.layout.recyclerview_item_repository, parent, false));
    }

    @Override
    public void onBindViewHolder(RepositoryViewHolder holder, int position) {
        holder.bindTo(repositories.get(position));
    }

    @Override
    public int getItemCount() {
        return repositories.size();
    }

    @Override
    public void accept(List<RepositoryViewModel> repositoryViewModels) throws Exception { // NOPMD
        repositories.clear();
        repositories.addAll(repositoryViewModels);
        notifyDataSetChanged();
    }

    public interface TrendingViewClickListener {
        void onRepositoryClick(RepositoryViewModel repository);
    }

    class RepositoryViewHolder extends RecyclerView.ViewHolder {
        private final RepositoryItemView repositoryItemView;

        RepositoryViewHolder(RepositoryItemView itemView) {
            super(itemView);
            repositoryItemView = itemView;
            repositoryItemView.setOnClickListener(v -> {
                RepositoryViewModel repository = repositories.get(getAdapterPosition());
                trendingViewClickListener.onRepositoryClick(repository);
            });
        }

        void bindTo(RepositoryViewModel viewModel) {
            repositoryItemView.bindTo(viewModel);
        }
    }
}

File: app/src/main/java/com/abishov/hexocat/home/repository/RepositoryAdapter.kt
package com.abishov.hexocat.home.repository

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.Adapter
import android.view.LayoutInflater
import android.view.ViewGroup
import com.abishov.hexocat.R
import com.abishov.hexocat.home.repository.RepositoryAdapter.RepositoryViewHolder
import io.reactivex.functions.Consumer
import java.util.*

class RepositoryAdapter(
  private val layoutInflater: LayoutInflater,
  private val trendingViewClickListener: TrendingViewClickListener,
  private val repositories: MutableList<RepositoryViewModel> = ArrayList()
) : Adapter<RepositoryViewHolder>(), Consumer<List<RepositoryViewModel>> {

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RepositoryViewHolder {
    return RepositoryViewHolder(
      layoutInflater.inflate(
        R.layout.recyclerview_item_repository, parent, false
      ) as RepositoryItemView
    )
  }

  override fun onBindViewHolder(holder: RepositoryViewHolder, position: Int) {
    holder.bindTo(repositories[position])
  }

  override fun getItemCount(): Int {
    return repositories.size
  }

  override fun accept(repositoryViewModels: List<RepositoryViewModel>) { // NOPMD
    repositories.clear()
    repositories.addAll(repositoryViewModels)
    notifyDataSetChanged()
  }

  interface TrendingViewClickListener {
    fun onRepositoryClick(repository: RepositoryViewModel)
  }

  inner class RepositoryViewHolder(private val repositoryItemView: RepositoryItemView) :
    RecyclerView.ViewHolder(repositoryItemView) {

    init {
      repositoryItemView.setOnClickListener {
        val repository = repositories[adapterPosition]
        trendingViewClickListener.onRepositoryClick(repository)
      }
    }

    fun bindTo(viewModel: RepositoryViewModel) {
      repositoryItemView.bindTo(viewModel)
    }
  }
}
